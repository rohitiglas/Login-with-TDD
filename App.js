import React, {useState} from 'react';
import {Alert, Button, TextInput, Text, View, StyleSheet} from 'react-native';

export const ValidationErrors = {
  FormEmpty: 'Form fields cannot be blank',
  UsernameEmpty: 'Username cannot be blank',
  PasswordEmpty: 'Password cannot be blank',
};
const App = () => {
  const [userName, setUserName] = useState('');
  const [password, setPassword] = useState('');
  const [validationError, setValidationError] = useState('');

  const onLogin = () => {
    if (userName.length === 0 && password.length === 0) {
      setValidationError(ValidationErrors.FormEmpty);
    } else if (userName.length === 0) {
      setValidationError(ValidationErrors.UsernameEmpty);
    } else if (password.length === 0) {
      setValidationError(ValidationErrors.PasswordEmpty);
    }
    else {
      Alert.alert('Credentials', `${userName} + ${password}`);
    }
  };
  return (
      <View style={styles.container}>
        {validationError.length !== 0 && (
            <Text testID={'text-error'} style={styles.errorText}>
              {validationError}
            </Text>
        )}

        <TextInput
            testID={'input-username'}
            value={userName}
            onChangeText={(text) => setUserName(text)}
            placeholder={'Username'}
            style={styles.input}
        />
        <TextInput
            testID={'input-password'}
            value={password}
            onChangeText={(text) => setPassword(text)}
            placeholder={'Password'}
            secureTextEntry={true}
            style={styles.input}
        />

        <Button
            testID={'submit-button'}
            title={'Login'}
            style={styles.input}
            onPress={onLogin}
        />
      </View>
  );
};

const styles = StyleSheet.create({
  container: {
    flex: 1,
    alignItems: 'center',
    justifyContent: 'center',
    backgroundColor: 'white',
  },
  errorText: {
    marginVertical: 10,
    color: 'red',
  },
  input: {
    width: 200,
    height: 44,
    padding: 10,
    borderWidth: 1,
    borderColor: 'black',
    marginBottom: 10,
  },
});

export default App;