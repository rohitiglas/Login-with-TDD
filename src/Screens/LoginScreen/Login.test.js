import 'react-native';
import React from 'react';
import App, {ValidationErrors} from './Login';
import {render, cleanup, fireEvent} from '@testing-library/react-native';
import {Alert} from "react-native";

jest.mock('react-native/Libraries/Alert/Alert', () => ({
    alert: jest.fn(),
}));

describe('Login Form', () => {
    let wrapper;
    // beforeEach() utility provided by Jest to render our wrapper before each test
    beforeEach(() => {
        wrapper = render(<App />);
    });
    // We also want each test to be isolated from each other —
    // thus, we need to perform un-mounting (“clean up”)
    // after each test. This can be done by using afterEach().
    afterEach(() => {
        cleanup();
        wrapper = null;
    });
    // Login Form Screen (i.e. the component) should display (render) correctly
    it('renders correctly', () => {
        expect(wrapper).toMatchSnapshot();
    });


//   Login Form should show validation error if -
//       a. user submits empty form (no username or password provided) ✅
//       b. user submits form with either username or password missing. ✅

    it("shows 'form-fields-empty' error if user submits without username or password", () => {
        const {getByTestId} = wrapper;
        const submitButton = getByTestId('submit-button');
        fireEvent(submitButton, 'onPress');
        const validationError = getByTestId('text-error');
        expect(validationError).toBeTruthy();
        expect(validationError.props.children)
            .toBe(ValidationErrors.FormEmpty);
    });

    it('shows error if user submits only a username', () => {
        const {getByTestId} = wrapper;
        const userNameText = getByTestId('input-username');
        const enteredUserName = 'RohitBansal';
        fireEvent(userNameText, 'onChangeText', enteredUserName);

        const submitButton = getByTestId('submit-button');
        fireEvent(submitButton, 'onPress');
        const validationError = getByTestId('text-error');

        expect(validationError).toBeTruthy();
        expect(validationError.props.children).toBe(ValidationErrors.PasswordEmpty);
    });

    it('shows error if user submits only a password', () => {
        const {getByTestId} = wrapper;
        const passwordText = getByTestId('input-password');
        const password = '12345678';
        fireEvent(passwordText, 'onChangeText', password);

        const submitButton = getByTestId('submit-button');
        fireEvent(submitButton, 'onPress');
        const validationError = getByTestId('text-error');

        expect(validationError).toBeTruthy();
        expect(validationError.props.children).toBe(ValidationErrors.UsernameEmpty);
    });


    it('should show Alert dialog if form validation successful', async () => {
        const {getByTestId} = wrapper;

        // Enter Username
        const userNameText = getByTestId('input-username');
        const enteredUserName = 'RohitBansal';
        fireEvent(userNameText, 'onChangeText', enteredUserName);

        // Enter Password
        const passwordText = getByTestId('input-password');
        const enteredPassword = '123456';
        fireEvent(passwordText, 'onChangeText', enteredPassword);

        // Press Submit Button
        const submitButton = getByTestId('submit-button');
        fireEvent.press(submitButton);

        // Check if Alert.alert() has been called
        const alertSpy = jest.spyOn(Alert, 'alert');
        expect(alertSpy).toHaveBeenCalled();
    });
});
