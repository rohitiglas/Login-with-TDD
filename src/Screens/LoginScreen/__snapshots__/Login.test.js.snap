// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login Form renders correctly 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "white",
      "flex": 1,
      "justifyContent": "center",
    }
  }
>
  <TextInput
    allowFontScaling={true}
    onChangeText={[Function]}
    placeholder="Username"
    rejectResponderTermination={true}
    style={
      Object {
        "borderColor": "black",
        "borderWidth": 1,
        "height": 44,
        "marginBottom": 10,
        "padding": 10,
        "width": 200,
      }
    }
    testID="input-username"
    underlineColorAndroid="transparent"
    value=""
  />
  <TextInput
    allowFontScaling={true}
    onChangeText={[Function]}
    placeholder="Password"
    rejectResponderTermination={true}
    secureTextEntry={true}
    style={
      Object {
        "borderColor": "black",
        "borderWidth": 1,
        "height": 44,
        "marginBottom": 10,
        "padding": 10,
        "width": 200,
      }
    }
    testID="input-password"
    underlineColorAndroid="transparent"
    value=""
  />
  <View
    accessibilityRole="button"
    accessibilityState={Object {}}
    accessible={true}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "opacity": 1,
      }
    }
    testID="submit-button"
  >
    <View
      style={
        Array [
          Object {},
        ]
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#007AFF",
              "fontSize": 18,
              "margin": 8,
              "textAlign": "center",
            },
          ]
        }
      >
        Login
      </Text>
    </View>
  </View>
</View>
`;
